src/connection.c

```diff
@@ -83,333 +82,460 @@ void set_global_reconnect_status(bool status)
+//--------------------------------------------------------------------
+// createNopollConnection_logic:
+
+// call stack:
+
+//  createNopollConnection
+//    find_servers
+//    keep_trying_connect     // keep trying till success or need to requery dns
+//      connect_and_wait      // tries both ipv6 and ipv4, if necessary
+//        nopoll_connect
+//        wait_connection_ready
+
+
+//--------------------------------------------------------------------
+/*  Defined in ParodusInternal.h
+#define SERVER_ADDR_LEN 256
+#define PORT_LEN 8
+
+typedef struct {
+  int allow_insecure;
+  char server_addr[SERVER_ADDR_LEN];
+  char port[PORT_LEN];
+} server_t;
+
+typedef struct {
+  server_t defaults;	// from command line
+  server_t jwt;		// from jwt endpoint claim
+  server_t redirect;	// from redirect response to
+			//  nopoll_conn_wait_until_connection_ready
+} server_list_t;
+*/
+
+void set_server_null (server_t *server)
+{
+  server->server_addr[0] = 0;
+}
+
+void set_server_list_null (server_list_t *server_list)
+{
+  set_server_null (&server_list->defaults);
+  set_server_null (&server_list->jwt);
+  set_server_null (&server_list->redirect);
+}
+
+int server_is_null (server_t *server)
+{
+  return (0 == server->server_addr[0]);
+}
+
+// If there's a redirect server, that's it,
+// else if there's a jwt server that's it,
+// else it's the default server
+
+server_t *get_current_server (server_list_t *server_list)
+{
+  if (!server_is_null (&server_list->redirect))
+    return &server_list->redirect;
+  if (!server_is_null (&server_list->jwt))
+    return &server_list->jwt;
+  return &server_list->defaults;
+}
+
+  
+int parse_server_url (const char *full_url, server_t *server)
+{
+  server->allow_insecure = parse_webpa_url (full_url,
+	server->server_addr, SERVER_ADDR_LEN, server->port, PORT_LEN);
+  return server->allow_insecure;
+}
+
+//--------------------------------------------------------------------
+/* --- Defined in ParodusInternal.h
+typedef struct {
+  int running;
+  struct timespec start_time;
+  struct timespec end_time;
+} expire_timer_t;
+*/
+
+void init_expire_timer (expire_timer_t *timer)
+{
+  timer->running = false;
+}
+
+int check_timer_expired (expire_timer_t *timer, long timeout_ms)
+{
+  long time_diff_ms;
+  
+  if (!timer->running) {
+    getCurrentTime(&timer->start_time);
+    timer->running = true;
+    ParodusInfo("First connect error occurred, initialized the connect error timer\n");
+    return false;
+  }
+  
+  getCurrentTime(&timer->end_time);
+  time_diff_ms = timeValDiff (&timer->start_time, &timer->end_time);
+  ParodusPrint("checking timeout difference:%ld\n", time_diff_ms);
+  if(time_diff_ms >= timeout_ms)
+    return true;
+  return false;
+}
+
+//--------------------------------------------------------------------
+/* ---- Defined in ParodusInternal.h
+typedef struct {
+  int max_delay;
+  int delay;
+} backoff_timer_t;
+*/
+
+void init_backoff_timer (backoff_timer_t *timer, int max_delay)
```

Comment:
For the backoff timer code, it probably makes sense to move this out & into it's own c/h file pair for simplicity.